pkg/storage/replica.go

```diff
@@ -980,6 +980,34 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
 				if !status.lease.OwnedBy(r.store.StoreID()) {
 					_, stillMember := r.mu.state.Desc.GetReplicaDescriptor(status.lease.Replica.StoreID)
 					if !stillMember {
+						// This would be the situation in which the lease holder gets removed when
+						// holding the lease, or in which a lease request erroneously gets accepted
+						// for a replica that is not in the replica set. Neither of the two can
+						// happen since appropriate mechanisms have been added:
+						//
+						// 1. Only the lease holder (at the time) schedules removal of a replica,
+						// but the lease can change hands and so the situation in which a follower
+						// coordinates a replica removal of the (new) lease holder is possible (if
+						// unlikely) in practice. In this situation, the new lease holder would at
+						// some point be asked to propose the replica change's EndTransaction to
+						// Raft. A check has been added that prevents proposals that amount to the
+						// removal of the proposer's (and hence lease holder's) Replica, preventing
+						// this scenario.
+						//
+						// 2. A lease is accepted for a Replica that has been removed. Without
+						// precautions, this could happen because lease requests are special in
+						// that they are the only command that is proposed on a follower (other
+						// commands may be proposed from followers, but not successfully so). For
+						// all proposals, processRaftCommand checks that their ProposalLease is
+						// compatible with the active lease for the log position. For commands
+						// proposed on the lease holder, the command queue then serializes
+						// everything. But lease commands get created on followers based on their
+						// local state and thus without being sequenced through the command queue.
+						// Thus a recently removed follower (unaware of its own removal) could
+						// submit a proposal for the lease (correctly using as a ProposerLease the
+						// last active lease), and would receive it given the up-to-date
+						// ProposerLease. Hence, an extra check is in order: processRaftCommand
+						// makes sure that leases for a replica not in the descriptor are bounced.
```

Comment:
I think this would be a little more clear if you replaced s/leases/lease requests/